### ISSUE TITLE
result_type doesn't take dtypes and doesn't match numpy

### ISSUE THREAD

**Issue by rgommers** [pytorch/pytorch#51284]
## 🚀 Feature

The signature for [torch.result_type](https://pytorch.org/docs/master/generated/torch.result_type.html) is currently:
```
result_type(tensor1, tensor2)
```
While [numpy.result_type](https://numpy.org/devdocs/reference/generated/numpy.result_type.html)'s signature is:
```
result_type(*arrays_and_dtypes)
```

It's not uncommon to mix tensors and dtypes to ensure you find both a dtype that input tensors can upcast to, and that's a minimum dtype in the dtype hierarchy, e.g.:
```
result_type(tensor1, tensor2, complex64)
```

Here is an example from scipy.signal: https://github.com/scipy/scipy/blob/5f4c4d802e5a56708d86909af6e5685cd95e6e66/scipy/signal/_upfirdn.py#L79

`torch.result_type` can be extended in a backwards compatible way to accepted a sequence of arrays and dtypes. I propose to do so.

## Motivation

1. It's useful
2. NumPy compatibility

## Additional details

It looks like this wasn't discussed in gh-26012, the PR that implemented `torch.result_type`.

**Comment by heitorschueroff** [comment]
I'm not sure we can offer the exact behavior as NumPy since C++ is strongly typed. Maybe we can do a manual dispatch from python API to convert the vararg into a list of tensor and a list of dtypes for the C++ API? @mruberry what do you think?

**Comment by mruberry** [comment]
This is a great idea. We should think about how best to implement it. If we have to we could implement the signature in Python and diverge the Python and C++ implementations.

**Comment by rgommers** [comment]
What I forgot to add to the motivation: the `result_type(tensor1, tensor2, min_dtype)` pattern wasn't all that useful without "mixed kind" casting, but now that we have integer to floating point casting which matches numpy, and (I think) float to complex as well, it's a lot more relevant.

**Comment by asi1024** [comment]
@mruberry @rgommers I think we need separate efforts to support these interfaces in Python and in C++.

- Python
  - Reveal `at::native::ResultTypeState` and `at::native::update_result_type_state` to Python.
  - Implement `torch.result_type` in Python and call the above functions from it.
- C++
  - Add new overload `update_result_type_state(const ScalarType&, const ResultTypeState&)` to accept dtype as an argument.
  - Support multi-length argument in `at::native::result_type` by variadic template.

If this is acceptable I would like to start implementing it 😃

**Comment by mruberry** [comment]
> @mruberry @rgommers I think we need separate efforts to support these interfaces in Python and in C++.
> 
> * Python
>   
>   * Reveal `at::native::ResultTypeState` and `at::native::update_result_type_state` to Python.
>   * Implement `torch.result_type` in Python and call the above functions from it.
> * C++
>   
>   * Add new overload `update_result_type_state(const ScalarType&, const ResultTypeState&)` to accept dtype as an argument.
>   * Support multi-length argument in `at::native::result_type` by variadic template.
> 
> If this is acceptable I would like to start implementing it 😃

This is an interesting approach that would reuse the existing C++ logic. It also sounds torturous, however. How feasible would it be to write this function in Python instead? Yes, that would be duplicative and diverge the Python and C++ API, but for compatibility we're principally interested in the Python API. We would also need to check if writing the operation in Python was supported by torch.fx and torch.jit scripting and tracing. My guess is that it will work with torch.fx and torch.jit tracing but not torch.jit scripting, but I could be mistaken.

**Comment by asi1024** [comment]
@mruberry
> How feasible would it be to write this function in Python instead? Yes, that would be duplicative and diverge the Python and C++ API, but for compatibility we're principally interested in the Python API.

I agree with that. How about re-implementing `at::native::update_result_type_state` in Python (I think it's quite easy: https://github.com/pytorch/pytorch/blob/4575028f6c593bd3f6ce9346882c1fb1808b18e1/aten/src/ATen/native/TypeProperties.cpp#L79-L113) and leaving C++ APIs as-is?

> My guess is that it will work with torch.fx and torch.jit tracing but not torch.jit scripting, but I could be mistaken.

I will check if TorchScript works after implementing it.

**Comment by mruberry** [comment]
@asi1024, cool! Let's try it. Let me know if you have additional questions.

**Comment by asi1024** [comment]
I re-implemented `at::native::update_result_type_state` as follows, but TorchScript does not work with this implementation.

https://gist.github.com/asi1024/ecf70644ee82aad38b1d024779335191

It seems that TorchScript cannot analyze multi-length arguments:

```py
import torch

def f(*args):
    return None

@torch.jit.script
def foo(x, y):
    f(x, y)
    return x + y
```

```
torch.jit.frontend.NotSupportedError: Compiled functions can't take variable number of arguments or use keyword-only arguments with defaults:
  File "run2.py", line 4
def f(*args):
      ~~~~~ <--- HERE
    return None
```

Ideas would be:
* Support variable arguments in TorchScript. I'm not sure how difficult this idea is, though.
* Do not support variable arguments in `result_type` for now (support only 2 arguments), and just add support for dtype argument by creating a 0-size dummy tensor from the given dtype and passing it to result_type as arguments.

What do you think? @mruberry

**Comment by mruberry** [comment]
> I re-implemented `at::native::update_result_type_state` as follows, but TorchScript does not work with this implementation.
> 
> https://gist.github.com/asi1024/ecf70644ee82aad38b1d024779335191
> 
> It seems that TorchScript cannot analyze multi-length arguments:
> 
> ```python
> import torch
> 
> def f(*args):
>     return None
> 
> @torch.jit.script
> def foo(x, y):
>     f(x, y)
>     return x + y
> ```
> 
> ```
> torch.jit.frontend.NotSupportedError: Compiled functions can't take variable number of arguments or use keyword-only arguments with defaults:
>   File "run2.py", line 4
> def f(*args):
>       ~~~~~ <--- HERE
>     return None
> ```

Thank you for investigating this option and verifying current JIT support, @asi1024. 


> Ideas would be:
> 
> * Support variable arguments in TorchScript. I'm not sure how difficult this idea is, though. or

My guess is that pursuing this would be very tricky.

> * Do not support variable arguments in `result_type` for now (support only 2 arguments), and just add support for dtype argument by creating a 0-size dummy tensor from the given dtype and passing it to result_type as arguments.

If we extend result_type in native_functions.yaml to take a tensor list (`Tensor[]`) then maybe we can do this. einsum() supports torchscript and varargs today:

```
https://github.com/pytorch/pytorch/blob/1145e2c6e238058d0f4a87e90db95c22a9dc2a87/torch/functional.py#L296
```

Maybe we could create a Python wrapper that performs the dtype->tensor conversion you're describing, then calls the updated result_type defined in native_functions.yaml with a list of tensors?

**Comment by asi1024** [comment]
@mruberry Sorry for my late reply :bow:

> If we extend result_type in native_functions.yaml to take a tensor list (`Tensor[]`) then maybe we can do this. einsum() supports torchscript and varargs today:
> 
> https://github.com/pytorch/pytorch/blob/1145e2c6e238058d0f4a87e90db95c22a9dc2a87/torch/functional.py#L296
> Maybe we could create a Python wrapper that performs the dtype->tensor conversion you're describing, then calls the updated result_type defined in native_functions.yaml with a list of tensors?

I found that `einsum` is not implemented as a Python wrapper but rather implemented as a function that accepts `Tensor[] tensors` in `native_functions.yaml`. However this approach cannot be used for `result_type` because both type and number of arguments cannot be statically declared.

I then found JIT-specific implementation for functions having variadic number of args and non-static types (e.g., `str.format(*args)`) are defined here: https://github.com/pytorch/pytorch/blob/v1.8.0/torch/csrc/jit/runtime/vararg_functions.cpp

I tried implementing [torch JIT runtime function for `result_type`](https://github.com/asi1024/pytorch/compare/71f4c5c1f436258adc303b710efb3f41b2d50c4e..465a38558e02288b89cb8cfa8ce55e4f9215f77b#diff-58586370e9b6ebf4c34f58954af43f5546359cfe188ee0d37a719c8c30753951R305-R332) in `vararg_functions.cpp`, and succeeded to support variadic & mixed type inputs of Tensors and dtypes (ScalarTypes). However, I couldn’t support scalar inputs because ScalarTypes and scalar values cannot be distinguished in JIT layer (issue #47964):

```py
import torch

def f():
    print(torch.float32)

f()  # prints "torch.float32"
torch.jit.script(f)()  # prints "6"
```

Now I think three different `result_type` implementations are needed.
1. New Python implementation of `result_type` for non-JIT runs, that accepts multiple Tensors, scalars or dtypes, i.e. compatible with the signature of NumPy and Data API.
2. Existing C++ implementations of `aten::result_type()` (declared in `native_functions.yaml`) for (Tensor, Tensor), (Tensor, Scalar), (Scalar, Tensor), (Scalar, Scalar) arguments. These implementations are for JIT runs and ATen C++ API users.
3. New "wildcard" C++ implementation (implemented in `vararg_functions.cpp`) for JIT runs with signatures not covered in 2.

The limitations in the wildcard implementation is that it does not support scalar inputs (every scalars are assumed to be a dtype (ScalarType)). This is the same behavior as `torch.view`
https://pytorch.org/docs/stable/tensors.html#torch.Tensor.view (which is blacklisted in JIT):

> view(dtype) -> Tensor
> This overload is not supported by TorchScript, and using it in a Torchscript program will cause undefined behavior.

```py
Traceback of TorchScript (most recent call last):
  File "test.py", line 11, in foo
    t = torch.tensor(10)
    print(t.view(torch.float32))
          ~~~~~~ <--- HERE
RuntimeError: shape '[6]' is invalid for input of size 1
```

**Comment by mruberry** [comment]
> I found that `einsum` is not implemented as a Python wrapper but rather implemented as a function that accepts `Tensor[] tensors` in `native_functions.yaml`. However this approach cannot be used for `result_type` because both type and number of arguments cannot be statically declared.

The Python definition for einsum is here:

https://github.com/pytorch/pytorch/blob/72ae924fad548769c0ba2e83c37b00963ca6bcb7/torch/functional.py#L158

And it has additional tests to validate it work with the JIT:

https://github.com/pytorch/pytorch/blob/4ca4640bae603d2b765a4063f6475d98f0dc218b/test/test_jit.py#L1584

Maybe taking a look at the wrapper, the native_functions.yaml implementation of einsum, and its tests would help?

> 1. New Python implementation of `result_type` for non-JIT runs, that accepts multiple Tensors, scalars or dtypes, i.e. compatible with the signature of NumPy and Data API.

This sounds like what we want. I think einsum was clever in how it manipulated its Python inputs into something C++ could handle. Maybe we could change the Python inputs to be a consistent type for varargs? Like maybe it can give C++ a list of dtypes and a list of integers, and the dtype is the dtype and the integers are an enum specifying the type of the object?

cc @heitorschueroff, who updated einsum

**Comment by asi1024** [comment]
@mruberry I'm sorry if my interpretation is wrong, but I think TorchScript does not refer to the definition in functional.py. For example, if I remove all arguments from the definition of `einsum` in functional.py, an error is raised when called einsum directly, but works correctly when called from the function decorated by `torch.jit.script`.

**Comment by mruberry** [comment]
> @mruberry I'm sorry if my interpretation is wrong, but I think TorchScript does not refer to the definition in functional.py. For example, if I remove all arguments from the definition of `einsum` in functional.py, an error is raised when called einsum directly, but works correctly when called from the function decorated by `torch.jit.script`.

Interesting! Let me ask one of our experts to give us a hand

**Comment by heitorschueroff** [comment]
@asi1024 you are correct in that TorchScript does not look at functional.py for the operator signatures. I followed your approach to add better JIT support for `einsum` here https://github.com/pytorch/pytorch/pull/59265. There is a way to tell TorchScript to use the overloads in functional.py as it's done for `tensordot` but that does not work for vararg parameters.

For result_type, I think the best solution at the moment is to add a function to functional.py that takes the vararg and groups them into a list of tensors, a list of scalars and a list of dtypes and calls a C++ function which we must defined in native_functions.yaml. Then also add the JIT function which similarly groups the arguments into lists and calls the same C++ function.

The problem with this is that, as @asi1024 noted, JIT cannot distinguish between scalar values and dtypes. One workaround would be to only accept tensors (including scalar tensors) and dtypes in the new vararg signatures. Another options is to simply wait until issue https://github.com/pytorch/pytorch/issues/47964 is addressed.

Since it is very common in NumPy to accept multiple types for the same parameter, it might be worthwhile to consider adding support for C++ 17 variant (c10::variant) in native_functions.yaml.

**Comment by heitorschueroff** [comment]
Also, the existing implementation of `result_type` differs from NumPy. NumPy's implementation inspects the actual value of the scalar to make sure that the value can be represented in the result dtype. For example...

```
>>> torch.result_type(torch.tensor([1], dtype=torch.float16), 1000000)
torch.float16
>>> np.result_type(np.array([1], dtype=np.float16), 1000000)
dtype('float64')
```

**Comment by mruberry** [comment]
> > NumPy's implementation inspects the actual value of the scalar to make sure that the value can be represented in the result dtype.
> 
> Please don't replicate that - that behavior is pretty awful; the array API standard doesn't have that, and NumPy itself is attempting to move away from value-based casting.

PyTorch also does not have this behavior.

**Comment by heitorschueroff** [comment]
To clarify, I think we should take the following approach.

1. Add a new overload in native_functions.yaml that accepts a list of tensors, a list of scalars and a list of dtypes
2. Add `result_type(*arrays_and_dtypes)` to functional.py. Group the arguments into lists of their respective types and call the new C++ function added in 1) 
3. Add JIT support for the vararg signature but only accept tensors for the moment, throw an error if the argument is not a Tensor
4. Once JIT can distinguish between Scalar and ScalarType, we can extend the JIT support for result_type.

**Comment by asi1024** [comment]
@heitorschueroff Thank you for your comment! It sounds good. I will work on it!

**Comment by tylerjereddy** [comment]
This bit me again while working on https://github.com/scipy/scipy/issues/18286 and https://github.com/tylerjereddy/scipy/pull/70 -- CuPy arrays and NumPy arrays worked, `torch` tensors could not be substituted directly and this was the first problem I encounted.

**Comment by hchau630** [comment]
any updates on this?
